src/kaleidoscope/plugin/Heatmap.h

```diff
@@ -34,16 +35,40 @@ class Heatmap : public LEDMode {
   EventHandlerResult onKeyswitchEvent(Key &mapped_key, byte row, byte col, uint8_t key_state);
   EventHandlerResult beforeEachCycle();
 
- protected:
-  void update(void) final;
+  // This class' instance has dynamic lifetime
+  //
+  class DynamicLEDMode : public LEDMode {
+   public:
 
- private:
-  static uint16_t heatmap_[ROWS][COLS];
-  static uint16_t highest_;
-  static uint32_t next_heatmap_comp_time_;
+    // Please note that storing the parent ptr is only required
+    // for those LED modes that require access to
+    // members of their parent class. Most LED modes can do without.
+    //
+    DynamicLEDMode(const Heatmap *parent);
 
-  static void shiftStats(void);
-  static cRGB computeColor(float v);
+    void resetMap();
+    EventHandlerResult onKeyswitchEvent(Key &mapped_key, byte row, byte col, uint8_t key_state);
+    EventHandlerResult beforeEachCycle();
```

Comment:
I'm curious about these hook functions in a dynamic LED mode. @noseglasses – how do they get called if the object doesn't always exist?

Reply:
By adding the `ACCESS_THIS_LED_MODE` macro to the parent plugin of a dynamic LED mode. The plugin is able to be informed about the mode id of its LED mode. It can then check if its LED mode is the current one. The heatmap plugin checks this in its handlers
```cpp
EventHandlerResult Heatmap::DynamicLEDMode::onKeyswitchEvent(Key &mapped_key, byte row, byte col, uint8_t key_state) {
   ...
   // Is my plugin the current one?
   if (::LEDControl.get_mode_index() != led_mode_id_)
      return EventHandlerResult::OK;
   ...
}
```
This would be more readable if `LEDControl`'s method would be named `get_current_mode_index` instead.

FWIW, every instance of a LED mode plugin (static and dynamic) is assigned an individual mode id (every solid color, e.g. has its own). This id is the corresponding index in the factory array.

Reply:
> how do they get called if the object doesn't always exist?

I didn't answer that part. The way I implemented it for some of the LED modes is to let the plugins handler check if its dynamic LED mode is current and then delegate the handler to the dynamic LED mode. For some the plugins, like the wavepool effect, this makes the handler code more readable as it is not cluttered with `dynamic_led_mode->...` statements. Those would be needed if the plugins handler would want to access members of its dynamic LED mode.

Reply:
This is the downside of using that macro; from the point of view of a plugin developer, `ACCESS_THIS_LED_MODE` is a mysterious black box, without much of a hint about where to find it. To me, it would help a great deal if this code was commented, with a reference to the file and line where the macro is defined, and a brief explanation of what it does.

Reducing the amount of boilerplate code is great, and providing a better guarantee that it's correct is even better, but needing to run a recursive `grep` in order to find all the methods and members belonging to a class whose header file I'm already looking at is a pretty big downside.

Reply:
`ACCESS_THIS_LED_MODE` is defined in header `LEDMode.h`. it works out of the box as every plugin that supplies LED modes no matter if static or dynamic, must include this header.

About the blackboxness of macros I totally agree with you. But I assume that most people that implement new LED modes will have a look on how the existing ones are done and just copy stuff from there. That's why I am less worried about the magic macro. It would of course be good to document it properly in those places where it is used right now.

Reply:
> The plugin is able to be informed about the mode id of its LED mode. It can then check if its LED mode is the current one. The heatmap plugin checks this in its handlers…

Thinking about this further, I think I can see an alternative, and I'd like to know what you think of the idea. Instead of passing the mode ID to the factory function when creating the dynamic LED mode, the constructor for the dynamic mode could use its "parent" pointer to call an "activate" function that sets a boolean so the parent knows its dynamic mode is active. Then, when it gets swapped out, the virtual destructor function could call "deactive" in the parent mode. The parent mode stores one byte either way (unless it combines the boolean into a bitfield), the factory function uses one fewer argument, and there's a vtable entry for the dynamic mode's destructor. I don't know how that compares, or if there's another reason this idea wouldn't work.

If this would work, and it's not worse in memory consumption, maybe it would be worth considering.

Reply:
I considered different approaches for the activity notification and went with the the least resource consuming one.

There are two problems with your proposed approach.
* Dynamic LED modes can but don't have to store a parent pointer. Some are completely independent from their parents (those that are not configurable). For those we can save two bytes.
* LED mode currently does not have a virtual destructor. If we would add it, we would add two bytes of PROGMEM for the additional entry in each LEDMode derived class' vtable. Also for those that have nothing particular to clean up.